aten/src/ATen/native/quantized/cpu/affine_quantizer.cpp

```diff
@@ -0,0 +1,236 @@
+    }
+  });
+}
+#endif  // USE_FBGEMM
+
+// TODO: add fbgemm for per channel
+void quantize_tensor_per_channel_affine_cpu(Tensor rtensor,
+                                          Tensor qtensor,
+                                          Tensor scales,
+                                          Tensor zero_points,
+                                          int64_t axis) {
+  AT_DISPATCH_QINT_TYPES(qtensor.scalar_type(), "quantize_tensor_per_channel_affine_cpu", [&]() {
+    int64_t batches = size_to_dim_(axis, rtensor.sizes());
+    int64_t elements_per_channel = size_from_dim_(axis + 1, rtensor.sizes());
+    int64_t channel = rtensor.size(axis);
+    auto scales_data = scales.data_ptr<double>();
+    auto zero_points_data = zero_points.data_ptr<int64_t>();
+    const float* rdata = rtensor.data_ptr<float>();
+    auto qdata = qtensor.data_ptr<scalar_t>();
+    for (auto b = 0; b < batches; ++b) {
+      for (auto c = 0; c < channel; ++c) {
+        for (auto e = 0; e < elements_per_channel; ++e) {
+          auto i = b * channel * elements_per_channel + c * elements_per_channel + e;
+          qdata[i] = quantize_val<scalar_t>(scales_data[c], zero_points_data[c], rdata[i]);
+        }
+      }
+    }
+  });
+}
+
+void dequantize_tensor_per_channel_affine_cpu(Tensor qtensor,
+                                            Tensor rtensor,
+                                            Tensor scales,
+                                            Tensor zero_points,
+                                            int64_t axis) {
+  AT_DISPATCH_QINT_TYPES(qtensor.scalar_type(), "dequantize_tensor_per_channel_affine_cpu", [&]() {
+    int64_t batches = size_to_dim_(axis, rtensor.sizes());
+    int64_t elements_per_channel = size_from_dim_(axis + 1, rtensor.sizes());
+    int64_t channel = rtensor.size(axis);
+    auto scales_data = scales.data_ptr<double>();
+    auto zero_points_data = zero_points.data_ptr<int64_t>();
+    const auto* qd = qtensor.data_ptr<scalar_t>();
+    float* rd = rtensor.data_ptr<float>();
+    for (auto b = 0; b < batches; ++b) {
+      for (auto c = 0; c < channel; ++c) {
+        for (auto e = 0; e < elements_per_channel; ++e) {
+          auto i = b * channel * elements_per_channel + c * elements_per_channel + e;
+          // We need to convert the qint8 value to float to ensure the subtraction
+          // subexpression returns a float
+          rd[i] = (static_cast<float>(qd[i].val_) - zero_points_data[c]) * scales_data[c];
+        }
+      }
+    }
+  });
+}
+} // anonymous namespace
+
+// Note: currently, quantize* functions are naively dispatched to a single realization. Move everything to the QuantizedOpKernels, once optimized versions are available.
+REGISTER_ARCH_DISPATCH(quantize_tensor_affine_stub, DEFAULT,  &quantize_tensor_affine_cpu);
+REGISTER_AVX_DISPATCH(quantize_tensor_affine_stub,  &quantize_tensor_affine_cpu);
+REGISTER_AVX2_DISPATCH(quantize_tensor_affine_stub,  &quantize_tensor_affine_cpu);
```

Comment:
Why are you calling these manually? Doesn't `REGISTER_DISPATCH` work?

Reply:
not sure, @z-a-f @dskhudia do you know?

Reply:
To give perspective onto this: implementation for the majority of the CPU quantized functions(perhaps, all) live in the [QuantizedOpKernels.cpp](https://github.com/alnfedorov/pytorch/blob/QuantizedCUDA/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp) file. This file is compiled multiple times with different vectorized instructions enabled(AVX/AVX2). Hence, `REGISTER_DISPATCH` in [QuantizedOpKernels.cpp](https://github.com/alnfedorov/pytorch/blob/QuantizedCUDA/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp) is called multiple times, each time with appropriate option substituted automatically(DEFAULT, AVX, AVX2). I might be mistaken somewhere, but the overall picture should stay the same.

I can move `[de]quantize_tensor*` functions to the [QuantizedOpKernels.cpp](https://github.com/alnfedorov/pytorch/blob/QuantizedCUDA/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp) to keep things simple. But there is the problem, `[de]quantize_tensor*` functions in the majority wouldn't benefit from this right now, yet it will increase compilation time and binary size. The default story(as far as I understand) is to relay on the fbgemm(or on the arm implementation with intrinsics) and fallback to the naive implementation only if optimized is not available. Naive implementations weren't compiled with AVX support before(and might not benefit from auto-vectorization, tests needed), so I decided to make a workaround until the better solution became clear.

Reply:
cc @jamesr66a @z-a-f @dskhudia could you comment on this one?

Reply:
I think it's fine to compile multiple times, also I feel we should have optimized implementations for quant/dequant functions on these platforms

Reply:
I agree with @jerryzh168, multiple compilation is fine, and it allows for the possibility of auto-vectorization, whereas otherwise we unconditionally don't have it. Op implementations like this are generally considered to be fine for biasing toward performance in the performance/binary size tradeoff

Reply:
> cc @jamesr66a @z-a-f @dskhudia @jianyuh could you comment on this one?

I think compiling multiple times is not a problem, and to be consistent I would stick with the `REGISTER_DISPATCH`.
@jamesr66a As for the optimizations, don't we care more about the binary size on mobile architectures?

Reply:
@z-a-f yeah but you're not compiling AVX* stuff for mobile :)

Reply:
> I think compiling multiple times is not a problem, and to be consistent I would stick with the REGISTER_DISPATCH
Agree with this.

Reply:
Alright, QuantizedOpKernels.cpp then.